File: bindings/julia/src/SafeFloat.jl
# SPDX-License-Identifier: PMPL-1.0
# SPDX-FileCopyrightText: 2025 Hyperpolymath

"""
    SafeFloat

Safe floating-point operations with NaN and infinity handling.
"""
module SafeFloat

export SafeF64
export safe_div, safe_sqrt, safe_log, safe_log10, safe_pow
export safe_sin, safe_cos, safe_tan, safe_asin, safe_acos, safe_atan
export is_safe, is_finite_value, clamp_finite
export approximately_equal, relative_eq
export EPSILON, MIN_POSITIVE, MAX_FINITE

"""Machine epsilon for Float64."""
const EPSILON = eps(Float64)

"""Smallest positive normalized Float64."""
const MIN_POSITIVE = floatmin(Float64)

"""Largest finite Float64."""
const MAX_FINITE = floatmax(Float64)

"""
    SafeF64

A floating-point wrapper that ensures values are always finite.
"""
struct SafeF64
    value::Float64

    function SafeF64(v::Float64)
        isfinite(v) || error("SafeF64 requires finite values")
        new(v)
    end
end

SafeF64(x::Real) = SafeF64(Float64(x))

"""
    is_safe(x::Float64) -> Bool

Check if a float is safe (finite).
"""
is_safe(x::Float64)::Bool = isfinite(x)

"""
    is_finite_value(x::Float64) -> Bool

Check if a float is finite (not NaN or Inf).
"""
is_finite_value(x::Float64)::Bool = isfinite(x)

"""
    clamp_finite(x::Float64) -> Float64

Clamp a value to finite range, converting NaN to 0 and Inf to max/min.
"""
function clamp_finite(x::Float64)::Float64
    if isnan(x)
        0.0
    elseif isinf(x)
        x > 0 ? MAX_FINITE : -MAX_FINITE
    else
        x
    end
end

"""
    safe_div(a::Float64, b::Float64) -> Union{Float64, Nothing}

Safely divide two floats. Returns nothing on division by zero or non-finite result.
"""
function safe_div(a::Float64, b::Float64)::Union{Float64, Nothing}
    b == 0.0 && return nothing
    result = a / b
    isfinite(result) ? result : nothing
end

"""
    safe_sqrt(x::Float64) -> Union{Float64, Nothing}

Safely compute square root. Returns nothing for negative numbers.
"""
function safe_sqrt(x::Float64)::Union{Float64, Nothing}
    x < 0.0 && return nothing
    result = sqrt(x)
    isfinite(result) ? result : nothing
end

"""
    safe_log(x::Float64) -> Union{Float64, Nothing}

Safely compute natural logarithm. Returns nothing for non-positive numbers.
"""
function safe_log(x::Float64)::Union{Float64, Nothing}
    x <= 0.0 && return nothing
    result = log(x)
    isfinite(result) ? result : nothing
end

"""
    safe_log10(x::Float64) -> Union{Float64, Nothing}

Safely compute base-10 logarithm. Returns nothing for non-positive numbers.
"""
function safe_log10(x::Float64)::Union{Float64, Nothing}
    x <= 0.0 && return nothing
    result = log10(x)
    isfinite(result) ? result : nothing
end

"""
    safe_pow(base::Float64, exp::Float64) -> Union{Float64, Nothing}

Safely compute power. Returns nothing on invalid or overflow.
"""
function safe_pow(base::Float64, exp::Float64)::Union{Float64, Nothing}
    # Handle special cases
    if base < 0.0 && !isinteger(exp)
        return nothing  # Complex result
    end
    if base == 0.0 && exp < 0.0
        return nothing  # Division by zero
    end

    result = base^exp
    isfinite(result) ? result : nothing
end

"""
    safe_sin(x::Float64) -> Union{Float64, Nothing}

Safely compute sine. Returns nothing for non-finite input.
"""
function safe_sin(x::Float64)::Union{Float64, Nothing}
    !isfinite(x) && return nothing
    sin(x)
end

"""
    safe_cos(x::Float64) -> Union{Float64, Nothing}

Safely compute cosine. Returns nothing for non-finite input.
"""
function safe_cos(x::Float64)::Union{Float64, Nothing}
    !isfinite(x) && return nothing
    cos(x)
end

"""
    safe_tan(x::Float64) -> Union{Float64, Nothing}

Safely compute tangent. Returns nothing for non-finite input or near singularity.
"""
function safe_tan(x::Float64)::Union{Float64, Nothing}
    !isfinite(x) && return nothing
    result = tan(x)
    isfinite(result) ? result : nothing
end

"""
    safe_asin(x::Float64) -> Union{Float64, Nothing}

Safely compute arcsine. Returns nothing for |x| > 1.
"""
function safe_asin(x::Float64)::Union{Float64, Nothing}
    abs(x) > 1.0 && return nothing
    asin(x)
end

"""
    safe_acos(x::Float64) -> Union{Float64, Nothing}

Safely compute arccosine. Returns nothing for |x| > 1.
"""
function safe_acos(x::Float64)::Union{Float64, Nothing}
    abs(x) > 1.0 && return nothing
    acos(x)
end

"""
    safe_atan(x::Float64) -> Union{Float64, Nothing}

Safely compute arctangent. Returns nothing for non-finite input.
"""
function safe_atan(x::Float64)::Union{Float64, Nothing}
    !isfinite(x) && return nothing
    atan(x)
end

"""
    approximately_equal(a::Float64, b::Float64; epsilon::Float64=EPSILON) -> Bool

Check if two floats are approximately equal within epsilon.
"""
function approximately_equal(a::Float64, b::Float64; epsilon::Float64=EPSILON)::Bool
    abs(a - b) <= epsilon
end

"""
    relative_eq(a::Float64, b::Float64; max_relative::Float64=EPSILON) -> Bool

Check if two floats are relatively equal.
"""
function relative_eq(a::Float64, b::Float64; max_relative::Float64=EPSILON)::Bool
    (!isfinite(a) || !isfinite(b)) && return false

    diff = abs(a - b)
    max_val = max(abs(a), abs(b))

    if max_val == 0.0
        diff == 0.0
    else
        diff / max_val <= max_relative
    end
end

# SafeF64 arithmetic
Base.:+(a::SafeF64, b::SafeF64) = SafeF64(clamp_finite(a.value + b.value))
Base.:-(a::SafeF64, b::SafeF64) = SafeF64(clamp_finite(a.value - b.value))
Base.:*(a::SafeF64, b::SafeF64) = SafeF64(clamp_finite(a.value * b.value))
Base.:/(a::SafeF64, b::SafeF64) = SafeF64(something(safe_div(a.value, b.value), 0.0))
Base.:-(a::SafeF64) = SafeF64(-a.value)

# Comparison
Base.:(==)(a::SafeF64, b::SafeF64) = a.value == b.value
Base.:<(a::SafeF64, b::SafeF64) = a.value < b.value
Base.:>(a::SafeF64, b::SafeF64) = a.value > b.value
Base.:<=(a::SafeF64, b::SafeF64) = a.value <= b.value
Base.:>=(a::SafeF64, b::SafeF64) = a.value >= b.value

# Conversion
Base.Float64(s::SafeF64) = s.value
Base.convert(::Type{Float64}, s::SafeF64) = s.value

# Display
Base.show(io::IO, s::SafeF64) = print(io, "SafeF64(", s.value, ")")

end # module

File: bindings/ocaml/lib/safe_float.ml
(* SPDX-License-Identifier: PMPL-1.0-or-later *)
(* SPDX-FileCopyrightText: 2025 Hyperpolymath *)

(** Safe floating-point operations with NaN/infinity handling. *)

type error =
  | Is_nan
  | Is_infinity
  | Division_by_zero
  | Overflow
  | Underflow

(** Check if value is a valid finite number. *)
let is_finite x =
  not (Float.is_nan x) && not (Float.is_inf x)

(** Safe add with overflow/NaN checking. *)
let safe_add a b =
  if Float.is_nan a || Float.is_nan b then Error Is_nan
  else
    let result = a +. b in
    if Float.is_nan result then Error Is_nan
    else if Float.is_inf result then Error Overflow
    else Ok result

(** Safe subtract with overflow/NaN checking. *)
let safe_sub a b =
  if Float.is_nan a || Float.is_nan b then Error Is_nan
  else
    let result = a -. b in
    if Float.is_nan result then Error Is_nan
    else if Float.is_inf result then Error Overflow
    else Ok result

(** Safe multiply with overflow/NaN checking. *)
let safe_mul a b =
  if Float.is_nan a || Float.is_nan b then Error Is_nan
  else
    let result = a *. b in
    if Float.is_nan result then Error Is_nan
    else if Float.is_inf result then Error Overflow
    else Ok result

(** Safe divide with zero/NaN checking. *)
let safe_div a b =
  if Float.is_nan a || Float.is_nan b then Error Is_nan
  else if b = 0.0 then Error Division_by_zero
  else
    let result = a /. b in
    if Float.is_nan result then Error Is_nan
    else if Float.is_inf result then Error Overflow
    else Ok result

(** Safe square root. *)
let safe_sqrt x =
  if Float.is_nan x then Error Is_nan
  else if x < 0.0 then Error Is_nan
  else Ok (Float.sqrt x)

(** Safe power. *)
let safe_pow base exp =
  if Float.is_nan base || Float.is_nan exp then Error Is_nan
  else
    let result = Float.pow base exp in
    if Float.is_nan result then Error Is_nan
    else if Float.is_inf result then Error Overflow
    else Ok result

(** Safe logarithm (natural). *)
let safe_log x =
  if Float.is_nan x then Error Is_nan
  else if x <= 0.0 then Error Is_nan
  else Ok (Float.log x)

(** Safe log base 10. *)
let safe_log10 x =
  if Float.is_nan x then Error Is_nan
  else if x <= 0.0 then Error Is_nan
  else Ok (Float.log10 x)

(** Safe exponential. *)
let safe_exp x =
  if Float.is_nan x then Error Is_nan
  else
    let result = Float.exp x in
    if Float.is_inf result then Error Overflow
    else Ok result

(** Clamp value to range. *)
let clamp ~min ~max x =
  if Float.is_nan x then Error Is_nan
  else if Float.is_nan min || Float.is_nan max then Error Is_nan
  else Ok (Float.max min (Float.min max x))

(** Check if two floats are approximately equal. *)
let approx_equal ?(epsilon=1e-10) a b =
  if Float.is_nan a || Float.is_nan b then false
  else
    let diff = Float.abs (a -. b) in
    diff <= epsilon || diff <= epsilon *. Float.max (Float.abs a) (Float.abs b)

(** Round to n decimal places. *)
let round_to_decimals n x =
  if Float.is_nan x then Error Is_nan
  else if Float.is_inf x then Error Is_infinity
  else
    let factor = Float.pow 10.0 (float_of_int n) in
    Ok (Float.round (x *. factor) /. factor)

(** Truncate to n decimal places. *)
let truncate_to_decimals n x =
  if Float.is_nan x then Error Is_nan
  else if Float.is_inf x then Error Is_infinity
  else
    let factor = Float.pow 10.0 (float_of_int n) in
    Ok (Float.trunc (x *. factor) /. factor)

(** Ceiling to n decimal places. *)
let ceil_to_decimals n x =
  if Float.is_nan x then Error Is_nan
  else if Float.is_inf x then Error Is_infinity
  else
    let factor = Float.pow 10.0 (float_of_int n) in
    Ok (Float.ceil (x *. factor) /. factor)

(** Floor to n decimal places. *)
let floor_to_decimals n x =
  if Float.is_nan x then Error Is_nan
  else if Float.is_inf x then Error Is_infinity
  else
    let factor = Float.pow 10.0 (float_of_int n) in
    Ok (Float.floor (x *. factor) /. factor)

(** Linear interpolation. *)
let lerp a b t =
  if Float.is_nan a || Float.is_nan b || Float.is_nan t then Error Is_nan
  else
    let result = a +. (b -. a) *. t in
    if Float.is_inf result then Error Overflow
    else Ok result

(** Map value from one range to another. *)
let map_range ~from_min ~from_max ~to_min ~to_max x =
  if Float.is_nan x || Float.is_nan from_min || Float.is_nan from_max ||
     Float.is_nan to_min || Float.is_nan to_max then
    Error Is_nan
  else if from_max = from_min then
    Error Division_by_zero
  else
    let normalized = (x -. from_min) /. (from_max -. from_min) in
    let result = to_min +. normalized *. (to_max -. to_min) in
    if Float.is_inf result then Error Overflow
    else Ok result

(** Safe modulo. *)
let safe_mod a b =
  if Float.is_nan a || Float.is_nan b then Error Is_nan
  else if b = 0.0 then Error Division_by_zero
  else Ok (Float.rem a b)

(** Convert to integer if it's a whole number. *)
let to_int_exact x =
  if Float.is_nan x then Error Is_nan
  else if Float.is_inf x then Error Is_infinity
  else if not (Float.is_integer x) then Error Is_nan
  else if x > float_of_int max_int || x < float_of_int min_int then Error Overflow
  else Ok (int_of_float x)

(** Format float with specific precision. *)
let format_precision n x =
  if Float.is_nan x then "NaN"
  else if Float.is_inf x then (if x > 0.0 then "Infinity" else "-Infinity")
  else Printf.sprintf "%.*f" n x

(** Parse float with validation. *)
let parse str =
  try
    let x = float_of_string str in
    if Float.is_nan x then Error Is_nan
    else if Float.is_inf x then Error Is_infinity
    else Ok x
  with _ -> Error Is_nan

(** Safe reciprocal (1/x). *)
let safe_reciprocal x =
  safe_div 1.0 x

(** Sign of a number (-1, 0, or 1). *)
let sign x =
  if Float.is_nan x then Error Is_nan
  else if x > 0.0 then Ok 1
  else if x < 0.0 then Ok (-1)
  else Ok 0

(** Hypotenuse (sqrt(a^2 + b^2)) with overflow protection. *)
let safe_hypot a b =
  if Float.is_nan a || Float.is_nan b then Error Is_nan
  else
    let result = Float.hypot a b in
    if Float.is_inf result then Error Overflow
    else Ok result

(** Safe arc tangent of y/x. *)
let safe_atan2 y x =
  if Float.is_nan y || Float.is_nan x then Error Is_nan
  else Ok (Float.atan2 y x)

(** Safe sine. *)
let safe_sin x =
  if Float.is_nan x then Error Is_nan
  else if Float.is_inf x then Error Is_infinity
  else Ok (Float.sin x)

(** Safe cosine. *)
let safe_cos x =
  if Float.is_nan x then Error Is_nan
  else if Float.is_inf x then Error Is_infinity
  else Ok (Float.cos x)

(** Safe tangent. *)
let safe_tan x =
  if Float.is_nan x then Error Is_nan
  else if Float.is_inf x then Error Is_infinity
  else
    let result = Float.tan x in
    if Float.is_inf result then Error Overflow
    else Ok result

(** Constants. *)
let pi = Float.pi
let e = Float.exp 1.0
let epsilon = Float.epsilon
let max_finite = Float.max_float
let min_positive = Float.min_float
